Akka.Serialization.MessagePack/Internal/IntIndexedMessagePackFormatterDict.cs

```diff
@@ -0,0 +1,80 @@
+using MessagePack.Formatters;
+
+namespace Akka.Serialization.MessagePack.Resolvers;
+
+/// <summary>
+/// A fast, Grow-only structure for holding Created formatters.
+/// Uses Int indexes per-type to avoid ConcurrentDictionary bucket costs,
+/// And avoids volatile/lock entirely on the happy path.
+/// </summary>
+/// <remarks>
+/// This can be (in some cases) less space efficient than a normal dictionary,
+/// However, for most internal uses, the cost is going to be fairly bounded,
+/// and <see cref="PolymorphicFormatterResolver"/>'s behavior results in
+/// <see cref="PolymorphicFormatterResolver.GetFormatter{T}"/> being a hot path,
+/// So in general the alloc cost is worth it. 
+/// </remarks>
+internal sealed class IntIndexedMessagePackFormatterDict
+{
+    private IMessagePackFormatter?[] _formatters;
+    private readonly object _lockObj = new();
+    public IntIndexedMessagePackFormatterDict(int initialSize)
+    {
+        _formatters = new IMessagePackFormatter[256];
+    }
+        
+    public bool TryGet(int i, out IMessagePackFormatter? formatter)
+    {
+        //No Fence here.
+        //If the read happened to be dirty, that's fine.
+        //Should it happen, it just means we trust `TryAdd`
+        //which does a full fence and then checks state.
+        //Shouldn't happen much and only on startup...
+        var f = _formatters;
```

Comment:
Meant to leave a comment about 'capturing' the array here, OTOH I left enough exposition where hopefully context is provided?